### ISSUE TITLE
irqbalance running in KVM

### ISSUE THREAD

**Issue by vcarceler** [Irqbalance/irqbalance#93]
Hi,

I just noticed that if you run ubuntu 18.04 server as guest in KVM (host is also ubuntu 18.04 server) with 2 vCPU. And you send a lot of network packets to the VM (for exemple using ping -f) then:

a) You get a 100% usage of just one core in the host machine.

b) In the guest machine there is only a core attending network interrupts.

Maybe I'm wrong but I guess that the expected behaviour is to use all the cores.

I just want to add that in the guest you may run top and doesn't show appreciable time in si state (time spent servicing software interrupts).

Thank you very much,
Victor

---

```
Package: irqbalance
Version: 1.3.0-0.1ubuntu0.18.04.1
```

```
usuario@ubuntu:~$ cat /proc/interrupts 
           CPU0       CPU1       
  0:          4          0   IO-APIC   2-edge      timer
  1:          0         17   IO-APIC   1-edge      i8042
  6:          0          3   IO-APIC   6-edge      floppy
  8:          0          0   IO-APIC   8-edge      rtc0
  9:          0          0   IO-APIC   9-fasteoi   acpi
 10:         22          0   IO-APIC  10-fasteoi   virtio3, ehci_hcd:usb1, uhci_hcd:usb2
 11:          0        135   IO-APIC  11-fasteoi   uhci_hcd:usb3, uhci_hcd:usb4, qxl
 12:        144          0   IO-APIC  12-edge      i8042
 14:         47        152   IO-APIC  14-edge      ata_piix
 15:          0          0   IO-APIC  15-edge      ata_piix
 24:          0          0   PCI-MSI 98304-edge      virtio1-config
 25:          4          0   PCI-MSI 98305-edge      virtio1-virtqueues
 26:          0          0   PCI-MSI 49152-edge      virtio0-config
 27:    1210604        193   PCI-MSI 49153-edge      virtio0-input.0
 28:          2         17   PCI-MSI 49154-edge      virtio0-output.0
 29:          0          0   PCI-MSI 114688-edge      virtio2-config
 30:       4188          0   PCI-MSI 114689-edge      virtio2-req.0
 31:          0         91   PCI-MSI 65536-edge      snd_hda_intel:card0
NMI:          0          0   Non-maskable interrupts
LOC:       3961       4134   Local timer interrupts
SPU:          0          0   Spurious interrupts
PMI:          0          0   Performance monitoring interrupts
IWI:          1          0   IRQ work interrupts
RTR:          0          0   APIC ICR read retries
RES:       2468       2631   Rescheduling interrupts
CAL:       2208       3045   Function call interrupts
TLB:          9         24   TLB shootdowns
TRM:          0          0   Thermal event interrupts
THR:          0          0   Threshold APIC interrupts
DFR:          0          0   Deferred Error APIC interrupts
MCE:          0          0   Machine check exceptions
MCP:          1          1   Machine check polls
HYP:          0          0   Hypervisor callback interrupts
ERR:          0
MIS:          0
PIN:          0          0   Posted-interrupt notification event
NPI:          0          0   Nested posted-interrupt event
PIW:          0          0   Posted-interrupt wakeup event
usuario@ubuntu:~$

```

```
root@ubuntu:~# irqbalance --debug
Isolated CPUs: 00000000
Adaptive-ticks CPUs: 00000000
Banned CPUs: 00000000
Package 1:  numa_node 0 cpu mask is 00000002 (load 0)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 0) 
                CPU number 1  numa_node is 0 (load 0)
Package 0:  numa_node 0 cpu mask is 00000001 (load 0)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 0) 
                CPU number 0  numa_node is 0 (load 0)
Adding IRQ 10 to database
Adding IRQ 31 to database
Adding IRQ 30 to database
Adding IRQ 29 to database
Adding IRQ 11 to database
Adding IRQ 9 to database
Adding IRQ 28 to database
Adding IRQ 26 to database
Adding IRQ 27 to database
Adding IRQ 25 to database
Adding IRQ 24 to database
Adding IRQ 0 to database
Adding IRQ 1 to database
Adding IRQ 6 to database
Adding IRQ 8 to database
Adding IRQ 12 to database
Adding IRQ 14 to database
Adding IRQ 15 to database
NUMA NODE NUMBER: -1
LOCAL CPU MASK: ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff

NUMA NODE NUMBER: 0
LOCAL CPU MASK: 00000003




-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 0)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 0) 
                CPU number 1  numa_node is 0 (load 0)
                  Interrupt 27 node_num is -1 (ethernet/0:325633) 
                  Interrupt 28 node_num is -1 (ethernet/0:5) 
                  Interrupt 30 node_num is -1 (storage/0:3) 
          Interrupt 11 node_num is -1 (legacy/0:60) 
  Interrupt 14 node_num is -1 (other/0:10) 
  Interrupt 8 node_num is -1 (other/0:0) 
  Interrupt 1 node_num is -1 (other/0:0) 
  Interrupt 24 node_num is -1 (other/0:0) 
  Interrupt 10 node_num is -1 (other/0:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 0)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 0) 
                CPU number 0  numa_node is 0 (load 0)
                  Interrupt 26 node_num is -1 (ethernet/0:0) 
                  Interrupt 29 node_num is -1 (storage/0:0) 
          Interrupt 9 node_num is -1 (legacy/0:0) 
  Interrupt 15 node_num is -1 (other/0:0) 
  Interrupt 12 node_num is -1 (other/0:0) 
  Interrupt 6 node_num is -1 (other/0:0) 
  Interrupt 0 node_num is -1 (other/0:0) 
  Interrupt 25 node_num is -1 (other/0:0) 
  Interrupt 31 node_num is -1 (other/0:0) 



-----------------------------------------------------------------------------
Selecting irq 28 for rebalancing
Selecting irq 14 for rebalancing
Package 1:  numa_node 0 cpu mask is 00000002 (load 240000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 240000000) 
                CPU number 1  numa_node is 0 (load 240000000)
                  Interrupt 27 node_num is -1 (ethernet/239918280:323340) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/46004:62) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 40010388)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 40003710) 
                CPU number 0  numa_node is 0 (load 40003710)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3710:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6678:9) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 300000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 300000000) 
                CPU number 1  numa_node is 0 (load 300000000)
                  Interrupt 27 node_num is -1 (ethernet/299852080:340741) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/63360:72) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 0)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 0) 
                CPU number 0  numa_node is 0 (load 0)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/1:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/1:9) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 250000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 250000000) 
                CPU number 1  numa_node is 0 (load 250000000)
                  Interrupt 27 node_num is -1 (ethernet/249827555:335339) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/52895:71) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 10000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 10000000) 
                CPU number 0  numa_node is 0 (load 10000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3333330:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6666660:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 250000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 250000000) 
                CPU number 1  numa_node is 0 (load 250000000)
                  Interrupt 27 node_num is -1 (ethernet/249752265:339799) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/52185:71) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 10000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 10000000) 
                CPU number 0  numa_node is 0 (load 10000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3333330:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6666660:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 250000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 250000000) 
                CPU number 1  numa_node is 0 (load 250000000)
                  Interrupt 27 node_num is -1 (ethernet/249944820:314396) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/54060:68) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 10000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 10000000) 
                CPU number 0  numa_node is 0 (load 10000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3333330:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6666660:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 260000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 260000000) 
                CPU number 1  numa_node is 0 (load 260000000)
                  Interrupt 27 node_num is -1 (ethernet/259743155:330883) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/59660:76) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 20000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 20000000) 
                CPU number 0  numa_node is 0 (load 20000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/6666665:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/13333330:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 290000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 290000000) 
                CPU number 1  numa_node is 0 (load 290000000)
                  Interrupt 27 node_num is -1 (ethernet/289917124:311404) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 11 node_num is -1 (legacy/68894:74) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 70000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 70000000) 
                CPU number 0  numa_node is 0 (load 70000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/23333330:5) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/46666660:10) 



-----------------------------------------------------------------------------
Selecting irq 30 for rebalancing
Package 1:  numa_node 0 cpu mask is 00000002 (load 240000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 240000000) 
                CPU number 1  numa_node is 0 (load 240000000)
                  Interrupt 27 node_num is -1 (ethernet/239808348:331227) 
          Interrupt 11 node_num is -1 (legacy/49232:68) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 10005068)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 10005068) 
                CPU number 0  numa_node is 0 (load 10005068)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3333330:5) 
                  Interrupt 30 node_num is -1 (storage/5068:7) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6666660:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 330000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 330000000) 
                CPU number 1  numa_node is 0 (load 330000000)
                  Interrupt 27 node_num is -1 (ethernet/329745924:330076) 
          Interrupt 11 node_num is -1 (legacy/53946:54) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 0)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 0) 
                CPU number 0  numa_node is 0 (load 0)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/1:5) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/1:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 290000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 290000000) 
                CPU number 1  numa_node is 0 (load 290000000)
                  Interrupt 27 node_num is -1 (ethernet/289723280:329231) 
          Interrupt 11 node_num is -1 (legacy/60720:69) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 10000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 10000000) 
                CPU number 0  numa_node is 0 (load 10000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3333330:5) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6666660:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 270000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 270000000) 
                CPU number 1  numa_node is 0 (load 270000000)
                  Interrupt 27 node_num is -1 (ethernet/269810592:333924) 
          Interrupt 11 node_num is -1 (legacy/49288:61) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 20000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 20000000) 
                CPU number 0  numa_node is 0 (load 20000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/5882350:5) 
                  Interrupt 30 node_num is -1 (storage/2352940:2) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/11764700:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 380000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 380000000) 
                CPU number 1  numa_node is 0 (load 380000000)
                  Interrupt 27 node_num is -1 (ethernet/379672096:320669) 
          Interrupt 11 node_num is -1 (legacy/95904:81) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 30000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 30000000) 
                CPU number 0  numa_node is 0 (load 30000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/10000000:5) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/20000000:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 290000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 290000000) 
                CPU number 1  numa_node is 0 (load 290000000)
                  Interrupt 27 node_num is -1 (ethernet/289881894:329786) 
          Interrupt 11 node_num is -1 (legacy/55377:63) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 30000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 30000000) 
                CPU number 0  numa_node is 0 (load 30000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/10000000:5) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/20000000:10) 



-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 330000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 330000000) 
                CPU number 1  numa_node is 0 (load 330000000)
                  Interrupt 27 node_num is -1 (ethernet/329733050:331390) 
          Interrupt 11 node_num is -1 (legacy/52735:53) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 40000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 40000000) 
                CPU number 0  numa_node is 0 (load 40000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/13333330:5) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/26666660:10) 




-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 00000002 (load 260000000)
        Cache domain 0:  numa_node is 0 cpu mask is 00000002  (load 260000000) 
                CPU number 1  numa_node is 0 (load 260000000)
                  Interrupt 27 node_num is -1 (ethernet/259667768:321371) 
          Interrupt 11 node_num is -1 (legacy/54136:67) 
  Interrupt 8 node_num is -1 (other/1:0) 
  Interrupt 1 node_num is -1 (other/1:0) 
  Interrupt 24 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
Package 0:  numa_node 0 cpu mask is 00000001 (load 10000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000001  (load 10000000) 
                CPU number 0  numa_node is 0 (load 10000000)
                  Interrupt 26 node_num is -1 (ethernet/1:0) 
                  Interrupt 29 node_num is -1 (storage/1:0) 
                  Interrupt 28 node_num is -1 (ethernet/3333330:5) 
                  Interrupt 30 node_num is -1 (storage/1:0) 
          Interrupt 9 node_num is -1 (legacy/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 12 node_num is -1 (other/1:0) 
  Interrupt 6 node_num is -1 (other/1:0) 
  Interrupt 0 node_num is -1 (other/1:0) 
  Interrupt 25 node_num is -1 (other/1:0) 
  Interrupt 31 node_num is -1 (other/1:0) 
  Interrupt 14 node_num is -1 (other/6666660:10) 
^Croot@ubuntu:~#

```

```
root@ubuntu:~# for i in $(seq 0 300); do grep . /proc/irq/$i/smp_affinity /dev/null 2>/dev/null; done
/proc/irq/0/smp_affinity:3
/proc/irq/1/smp_affinity:2
/proc/irq/2/smp_affinity:3
/proc/irq/3/smp_affinity:3
/proc/irq/4/smp_affinity:3
/proc/irq/5/smp_affinity:3
/proc/irq/6/smp_affinity:1
/proc/irq/7/smp_affinity:3
/proc/irq/8/smp_affinity:2
/proc/irq/9/smp_affinity:1
/proc/irq/10/smp_affinity:2
/proc/irq/11/smp_affinity:2
/proc/irq/12/smp_affinity:1
/proc/irq/13/smp_affinity:3
/proc/irq/14/smp_affinity:1
/proc/irq/15/smp_affinity:1
/proc/irq/24/smp_affinity:2
/proc/irq/25/smp_affinity:1
/proc/irq/26/smp_affinity:1
/proc/irq/27/smp_affinity:2
/proc/irq/28/smp_affinity:1
/proc/irq/29/smp_affinity:3
/proc/irq/30/smp_affinity:3
/proc/irq/31/smp_affinity:1
root@ubuntu:~#
```

```
root@ubuntu:~# lstopo-no-graphics
Machine (985MB)
  Package L#0 + L3 L#0 (16MB) + L2 L#0 (4096KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0 + PU L#0 (P#0)
  Package L#1 + L3 L#1 (16MB) + L2 L#1 (4096KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1 + PU L#1 (P#1)
  HostBridge L#0
    PCI 8086:7010
      Block(Removable Media Device) L#0 "sr0"
    PCI 1b36:0100
      GPU L#1 "controlD64"
      GPU L#2 "card0"
    PCI 1af4:1000
    PCI 1af4:1001
root@ubuntu:~#
```

**Comment by nhorman** [comment]
It would appear from your irqbalance output that irqbalance is working properly, in that is has assigned each receive queue for each ethernet device to a unique cpu.  It does this to maximize the instruction and data cache hit rates for each irq.

you are only seeing irq traffic on one cpu because you are flooding a single network queue/interface with traffic, and each queue has only one unique interrupt.  because irqs are self serialized (that is to say, a single irq is only processed by one cpu at one time).  because of that, even though it may be comforting to see a single irq have its trigger counts increase on multiple cpus (giving the sense of parallelism), thats not as efficient, because each of those triggers has to be handled serially, and doing so on multiple cpus reduces your cache hit rate.

If you want increased performance by receiving frames on multiple cpus, you need to add queues to your network device (allowing for the use of multiple msi irqs which can be handled in parallel).  To do this you either need to configure that queue count in your host (if its supported by your hardware), or change your hardware to a multiqueue device and pass it into the guest via sriov/vfio.
